Add support for building with Zig

plajjan opened the pull request:
This adds a build.zig that allows building the garbage collector using the Zig build system. It implements a subset of the configuration options offered by CMake and following the naming convention in CMake for all those options.

There is a new GitHub Actions configuration file that runs a job that uses zig to compile the collector and targetting different targets using cross-compilation. It doesn't run any actual tests (and it should!).

The README is slightly updated, not just adding an explanation of zig but restructuring the whole section on how to build the GC somewhat. I moved mentions of libatomic_ops to a separate heading. All? modern compilers include atomic intrinsics so it feels like we don't have to include mentions of libatomic_ops in the default "how to build this" section but can mention it a little later. I tried to make it easier to read and more direct and to the point for new users. Let me know if you don't agree with this restructuring and I'll change back. I think there's more stuff to reword in the README but I don't want to focus too much on that in this PR since it's really about introducing zig.

The build.zig follows the format used in zig v0.12. The final 0.12 version is not out yet, so users will have to use a build from the zig master branch for now. This is not ideal but the whole zig build system is so new that there have been backwards incompatible changes from 0.11 which is why we opt to use 0.12 that is likely going to be closer to the final form.

plajjan commented:
@ivmai here's a first draft of adding in support for the zig build system. Please let me know what you think!? :)

plajjan commented:
I mean, there are things that are missing, like building dynamic libraries so I don't consider this ready for merge. I just wanted to get the current version out there to get feedback on the overall look and feel of it!

ivmai commented on .github/workflows/zig-build.yml:
```diff
@@ -0,0 +1,29 @@
+    - uses: actions/checkout@v4
+    - name: "Install zig"
+      run: |
+        curl -o zig.tar.xz https://ziglang.org/builds/zig-linux-x86_64-0.12.0-dev.1814+5c0d58b71.tar.xz
```
Can we do w/o an intermediate (zig.tar.xz) file? Like this (mind '-' symbol):
```
mkdir zig
wget -O - <url> | tar xf - --xz --strip-components=1 --directory zig
```

plajjan replied:
Yes, fixed!

ivmai commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+        default_enable_threads = false;
+    }
+    const enable_threads = b.option(bool, "enable_threads", "Support threads") orelse default_enable_threads;
+    const enable_parallel_mark = b.option(bool, "enable_parallel_mark", "Parallelize marking and free list construction") orelse true;
```
Would be nice to reformat these lines to have <80 chars on each line. (Yes, the project still uses this limitation, I'm thinking of a global reformat, but not now.)

plajjan replied:
Fixed!

ivmai commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    var default_enable_threads = true;
+    if (t.isWasm()) { // matches both emscripten and wasi
+        default_enable_threads = false;
+    }
```
We need an option to choose between static and shared library builds. Default is the latter one.
E.g. BUILD_SHARED_LIBS (or enable_shared).
It is fine for this PR to fail build if shared lib build is unsupported for now.

plajjan replied:
I have added in support for shared libs now!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+        "-DNO_EXECUTE_PERMISSION",
+        // Zig comes with clang that supports atomics so we can hardcode this
+        "-DGC_BUILTIN_ATOMIC",
+        // TODO: Should this be configurable?
```
Right, I think this is carried over from the Acton use. I can't quite remember why this was configured like this. I've skimmed through some of the documentation to get a better idea of what USE_PROC_FOR_LIBRARIES actually does.

gcconfig.h currently sets USE_PROC_FOR_LIBRARIES on Linux with threads and with REDIRECT_MALLOC enabled.

Acton uses this combination (linux + thread + redirect), so this is relevant. For some reason I have disabled USE_PROC_FOR_LIBRARIES but I don't know exactly why. Doc says it is slow. It also prints a warning. Is it always necessary to do this or are there certain conditions under which it is required?

If it is always required for this combination of platform / features, then I think we need an option to disable the warning.

If it is not always required, i.e. it could be configured by the user, then we need a config option for it, e.g. exposing `NO_PROC_FOR_LIBRARIES` to the end user.

ivmai replied:
I think we could just remove the warning, it works better (more stable) after certain fixes.

plajjan replied:
Ok!

plajjan replied:
I've removed the `NO_PROC_FOR_LIBRARIES` from this PR which I think is the most important part. I'd rather look into the warning in a separate PR if we want to remove it...

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    }) catch unreachable;
+
+    if (enable_threads) {
+        flags.appendSlice(&.{"-DGC_THREADS"}) catch unreachable;
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+        } else {
+            lib.addCSourceFiles(.{
+                .files = &.{
+                    "gc_dlopen.c",
```
Fixed!

plajjan replied:
Zig is lacking some headers and stuff to make it work out of the box but I think it's possible to use if you supply your own cygwin threads library / headers etc. I saw an issue at some point about someone working on it, so it might be fixed. Anyway, it's aligned now!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+        flags.appendSlice(&.{"-DGC_THREADS"}) catch unreachable;
+        if (enable_parallel_mark)
+            flags.appendSlice(&.{"-DPARALLEL_MARK"}) catch unreachable;
+        if (enable_thread_local_alloc) {
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    });
+    lib.addIncludePath(.{ .path = "include" });
+    lib.linkLibC();
+    lib.installHeader("include/gc.h", "gc.h");
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    flags.appendSlice(&.{
+        "-fno-sanitize=undefined",
+        // Always enabled
+        "-DALL_INTERIOR_POINTERS",
```
Sort of fixed. I switched away from manipulating raw flags to instead use a function, `lib.defineCMacro()` which I think looks cleaner (and is most certainly the intended to be used for this). That doesn't take `-D` at all..

plajjan replied:
I switched back from using defineCMacro because it is easier to reuse the flags for multiple compilation targets if I don't add them directly to the lib but keep them as a separate list. I needed this to apply the same set of flags for tests as well, thus your original point stands. I have added the space!

plajjan commented on .github/workflows/zig-build.yml:
```diff
@@ -0,0 +1,29 @@
+      matrix:
+        ttriple: [ aarch64-linux-musl, wasm32-wasi, x86_64-linux-gnu.2.27, x86_64-linux-musl, x86_64-windows-gnu ]
+
+    steps:
```
Fixed

plajjan commented on .github/workflows/zig-build.yml:
```diff
@@ -0,0 +1,29 @@
+# This workflow uses Zig and its excellent cross-compilation support to test
+# compiling for multiple platforms. No tests are actually run since it would
+# require emulation. We should add a test for the native platform though.
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    });
+    lib.addIncludePath(.{ .path = "include" });
+    lib.linkLibC();
+    lib.installHeader("include/gc.h", "gc.h");
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+        flags.appendSlice(&.{"-DUSE_MMAP"}) catch unreachable;
+    }
+
+    if (enable_large_config)
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+        flags.appendSlice(&.{"-DLARGE_CONFIG"}) catch unreachable;
+
+    const source_files = [_][]const u8{
+        "alloc.c",
```
Ok, I actually did that originally but then reordered it to follow the source file order in CMake to make it easier to compare between the two. Maybe sort source file list CMake too?

ivmai replied:
Yes, reorder in cmake script (and probably in makefile.am)

plajjan replied:
Ok, fixed!

plajjan reviewed (commented):
Addressed most comments so far.

plajjan commented:
@ivmai so I've added in compilation and running of the basic gctest as well. It works when building a static lib and disabling threads:
```sh
kll@Boxy:~/terastream/acton-deps/bdwgc$ ~/terastream/acton/dist/zig/zig build -Dbuild_tests -Dbuild_shared_libs=false -Denable_threads=false test
Switched to incremental mode
Reading dirty bits from /proc
Completed 6 tests
Allocated 4154109 collectable objects
Allocated 1218 uncollectable objects
Allocated 8219232 atomic objects
Reallocated 756 objects
Garbage collection after fork is tested too
Finalized 13216/13216 objects - finalization is probably OK
Total number of bytes allocated is 1143245488
Total memory use by allocated blocks is 638976 bytes
Final heap size is 172032 bytes
Obtained 12488704 bytes from OS
Final number of reachable objects is 2423
Full/world-stopped collections took 23/484 ms
Completed 2475 collections
Collector appears to work
kll@Boxy:~/terastream/acton-deps/bdwgc$ 
```

If I enable threads, I get a failure:
```sh
kll@Boxy:~/terastream/acton-deps/bdwgc$ ~/terastream/acton/dist/zig/zig build -Dbuild_tests -Dbuild_shared_libs=false -Denable_threads=true test
Switched to incremental mode
Reading dirty bits from /proc
GC_is_visible produced wrong failure indication
Test failed
run gctest: error: the following command terminated unexpectedly:
/home/kll/terastream/acton-deps/bdwgc/zig-cache/o/71dd4f8899463605f4330d1f46afd9e6/gctest 
Build Summary: 19/21 steps succeeded; 1 failed (disable with --summary none)
test transitive failure
└─ run gctest failure
error: the following build command failed with exit code 1:
/home/kll/terastream/acton-deps/bdwgc/zig-cache/o/6ca84adc5567722c0e59ba2f7e1f9200/build /home/kll/terastream/acton/dist/zig/zig /home/kll/terastream/acton-deps/bdwgc /home/kll/terastream/acton-deps/bdwgc/zig-cache /home/kll/.cache/zig --seed 0x2b9282f8 -Dbuild_tests -Dbuild_shared_libs=false -Denable_threads=true test
kll@Boxy:~/terastream/acton-deps/bdwgc$
```

Similarly when building a shared lib also without threads, I get a different error:
```sh
kll@Boxy:~/terastream/acton-deps/bdwgc$ ~/terastream/acton/dist/zig/zig build -Dbuild_tests -Dbuild_shared_libs=true -Denable_threads=false test
Switched to incremental mode
Reading dirty bits from /proc
List reversal produced incorrect list - collector is broken
Test failed
run gctest: error: the following command terminated unexpectedly:
/home/kll/terastream/acton-deps/bdwgc/zig-cache/o/9285cc5d5ea49b5fec8b8149ebdb78e2/gctest 
Build Summary: 19/21 steps succeeded; 1 failed (disable with --summary none)
test transitive failure
└─ run gctest failure
error: the following build command failed with exit code 1:
/home/kll/terastream/acton-deps/bdwgc/zig-cache/o/6ca84adc5567722c0e59ba2f7e1f9200/build /home/kll/terastream/acton/dist/zig/zig /home/kll/terastream/acton-deps/bdwgc /home/kll/terastream/acton-deps/bdwgc/zig-cache /home/kll/.cache/zig --seed 0x72fd843e -Dbuild_tests -Dbuild_shared_libs=true -Denable_threads=false test
kll@Boxy:~/terastream/acton-deps/bdwgc$ 
```

When building with the equivalent options for CMake, the tests pass just fine. I've been trying to pin down where the compilation differs but haven't found anything so far. It could be that it's a difference due to zig default clang flags. I wonder if the error messages can guide to the error but I don't quite understand them, in particular the "GC_is_visible produced wrong failure indication". Could you possibly point me in the general direction where I should dig deeper?

I just pushed the latest code that fails as described above.

plajjan commented:
Nevermind, I found the cause. The flags were only applied to compilation of the lib, not of gctest itself and that discrepancy caused the failures. Having fixed that, I seem to be having some success with the tests. More to be reported!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    const enable_disclaim = b.option(bool, "enable_disclaim", "Support alternative finalization interface") orelse true;
+    const enable_large_config = b.option(bool, "enable_large_config", "Optimize for large heap or root set") orelse true;
+    const enable_mmap = b.option(bool, "enable_mmap", "Use mmap instead of sbrk to expand the heap") orelse true;
+    const enable_munmap = b.option(bool, "enable_munmap", "Return page to the OS if empty for N collections") orelse true;
```
I have now added all options to the argument parser, some commented out which we can implement at a later point in time.

ivmai commented on build.zig:
```diff
@@ -0,0 +1,383 @@
+    lib.linkLibC();
+    if (install_headers) {
+        lib.installHeader("include/gc.h", "gc.h");
+        lib.installHeader("include/gc/cord.h", "gc/cord.h");
```
is it possible to make some function for convenience (and reduce possibility of a bug) so that to source specify file name and based destination folder, e.g. installHeaderTo("gc/cord.h", "include") ?

plajjan replied:
Yes, I just added in our own `installHeader` function that does this now so we don't have to repeat the filename in source and destination.

FYI, the first argument is the source, so it's `include/gc.h` in the source git repo. The second argument is the destination and it is implied that it always ends up in the install prefix directory under `include/`. The prefix dirs can be modified with command line args to `zig build`, like so:
```
zig build --help

...

General Options:
  -p, --prefix [path]          Override default install prefix
  --prefix-lib-dir [path]      Override default library directory path
  --prefix-exe-dir [path]      Override default executable directory path
  --prefix-include-dir [path]  Override default include directory path
```

ivmai commented on build.zig:
```diff
@@ -0,0 +1,383 @@
+    const enable_cplusplus = b.option(
+        bool, "enable_cplusplus", "C++ support") orelse false;
+    const build_shared_libs = b.option(
+        bool, "build_shared_libs",
```
I think it is better change name to all caps (BUILD_SHARED_LIBS) as in cmake script

plajjan replied:
Fixed!

ivmai commented on build.zig:
```diff
@@ -0,0 +1,383 @@
+    }
+
+    if (enable_redirect_malloc) {
+        lib.defineCMacro("REDIRECT_MALLOC=GC_malloc", "1");
```
I guess it should be: lib.defineCMacro("REDIRECT_MALLOC", "GC_malloc")

plajjan replied:
Yeah that was wrong, but I'm changing back to the use of flags so it's correct again :)

plajjan commented on build.zig:
```diff
@@ -0,0 +1,175 @@
+    defer flags.deinit();
+
+    if (t.abi.isMusl()) {
+        std.log.debug("Musl detected", .{});
```
Fixed!

plajjan commented on build.zig:
```diff
@@ -0,0 +1,383 @@
+        }
+    }
+
+    if (build_shared_libs and !(target.isDarwin() or target.isDragonFlyBSD() or target.isFreeBSD() or target.isNetBSD() or target.isOpenBSD())) {
```
Fixed!

ivmai commented:
I am reviewing the patch, I will commit some changes tomorrow.

plajjan commented:
@ivmai I'm seeing some intermittent test failures, like in https://github.com/ivmai/bdwgc/actions/runs/7273056790 where 1 out of 12 jobs failed. I can reproduce it locally where it is intermittent. I think it only happens with assertions enabled.

For example, here are two runs, one succeeding and one failing:
```
kll@ThinkYoga:~/terastream/acton-deps/bdwgc$ /home/kll/terastream/acton-deps/bdwgc/zig-cache/o/dc24990df2d413d251b3167248ce0c18/gctest
Switched to incremental mode
Reading dirty bits from /proc
Completed 6 tests
Allocated 4281275 collectable objects
Allocated 1218 uncollectable objects
Allocated 8216838 atomic objects
Reallocated 756 objects
Garbage collection after fork is tested too
Finalized 13218/13218 objects - finalization is probably OK
Total number of bytes allocated is 1140762464
Total memory use by allocated blocks is 806912 bytes
Final heap size is 217088 bytes
Obtained 13279232 bytes from OS (of which 11415552 bytes unmapped)
Final number of reachable objects is 2434
Full/world-stopped collections took 215/4224 ms
Completed 2392 collections
Collector appears to work
kll@ThinkYoga:~/terastream/acton-deps/bdwgc$ /home/kll/terastream/acton-deps/bdwgc/zig-cache/o/dc24990df2d413d251b3167248ce0c18/gctest
Switched to incremental mode
Reading dirty bits from /proc
List reversal produced incorrect list - collector is broken
Test failed
Aborted (core dumped)
```

```
kll@ThinkYoga:~/terastream/acton-deps/bdwgc$ gdb /home/kll/terastream/acton-deps/bdwgc/zig-cache/o/dc24990df2d413d251b3167248ce0c18/gctest core
...
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007f57740e1d2f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007f5774092ef2 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f577407d472 in __GI_abort () at ./stdlib/abort.c:79
#4  0x00000000002090b6 in check_ints (list=0x7f577401c920, low=1, up=49) at tests/gctest.c:573
#5  0x0000000000208c63 in reverse_test_inner (data=0x1) at tests/gctest.c:936
#6  0x00007f5774286fb0 in GC_call_with_gc_active (fn=0x208970 <reverse_test_inner>, client_data=0x1) at extra/../misc.c:2284
#7  0x000000000020899b in reverse_test_inner (data=0x0) at tests/gctest.c:860
#8  0x00007f5774287185 in GC_do_blocking_inner (data=0x7fffe58063a0 "p\211 ", context=0x7fffe5805f80) at extra/../misc.c:2318
#9  0x00007f57742828e4 in GC_with_callee_saves_pushed (fn=0x7f57742870a0 <GC_do_blocking_inner>, arg=0x7fffe58063a0 "p\211 ") at extra/../mach_dep.c:337
#10 0x00007f5774287420 in GC_do_blocking (fn=0x208970 <reverse_test_inner>, client_data=0x0) at extra/../misc.c:2366
#11 0x00000000002080a4 in reverse_test () at tests/gctest.c:996
#12 0x00000000002066e1 in run_one_test () at tests/gctest.c:1911
#13 0x00000000002052c8 in main () at tests/gctest.c:2325
(gdb) 
```

<details>
  <summary>Click me</summary>
  
```
(gdb) bt full
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
        tid = <optimized out>
        ret = 0
        pd = <optimized out>
        old_mask = {__val = {140737043782864}}
        ret = <optimized out>
#1  0x00007f57740e1d2f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
No locals.
#2  0x00007f5774092ef2 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
        ret = <optimized out>
#3  0x00007f577407d472 in __GI_abort () at ./stdlib/abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x20, sa_sigaction = 0x20}, sa_mask = {__val = {140013585154752, 140737043782960, 2134287, 25769803826, 140013585154720, 140737043782992, 2134287, 21474836530, 
              140013585653813, 140737043783024, 140013584338480, 140737043783088, 140013584338560, 140013584338672, 140013584338512, 140013584338528}}, sa_flags = 1948806097, sa_restorer = 0x7fffe5805db0}
#4  0x00000000002090b6 in check_ints (list=0x7f577401c920, low=1, up=49) at tests/gctest.c:573
No locals.
#5  0x0000000000208c63 in reverse_test_inner (data=0x1) at tests/gctest.c:936
        i = 10
        b = 0x7f577401c200
        c = 0x7f5773f52da8
        d = 0x7f5774025d18
        e = 0x7f5774025d00
        f = 0x7f5774028fc0
        g = 0x7f5774022000
        h = 0x7f5773c14000
#6  0x00007f5774286fb0 in GC_call_with_gc_active (fn=0x208970 <reverse_test_inner>, client_data=0x1) at extra/../misc.c:2284
        stacksect = {saved_stack_ptr = 0x7fffe5805f48 "\240c\200\345\377\177", prev = 0x0}
#7  0x000000000020899b in reverse_test_inner (data=0x0) at tests/gctest.c:860
        i = 0
        b = 0x0
        c = 0x0
        d = 0x0
        e = 0x0
        f = 0x0
        g = 0x0
        h = 0x0
#8  0x00007f5774287185 in GC_do_blocking_inner (data=0x7fffe58063a0 "p\211 ", context=0x7fffe5805f80) at extra/../misc.c:2318
        d = 0x7fffe58063a0
#9  0x00007f57742828e4 in GC_with_callee_saves_pushed (fn=0x7f57742870a0 <GC_do_blocking_inner>, arg=0x7fffe58063a0 "p\211 ") at extra/../mach_dep.c:337
        getcontext_works = 1 '\001'
        dummy = 0
        context = 0x7fffe5805f80 ""
        ctxt = {uc_flags = 0, uc_link = 0x0, uc_stack = {ss_sp = 0x0, ss_flags = 0, ss_size = 0}, uc_mcontext = {gregs = {0, 0, 0, 0, 140013584338560, 140013584338672, 140013584338512, 140013584338528, 140737043783552, 
              140737043784608, 140737043784592, 140013584338480, 0, 0, 477568, 140737043783536, 140013587671079, 0, 0, 0, 0, 0, 0}, fpregs = 0x7fffe5806128, __reserved1 = {0, 0, 0, 0, 0, 0, 0, 0}}, uc_sigmask = {__val = {
              0 <repeats 16 times>}}, __fpregs_mem = {cwd = 895, swd = 65535, ftw = 0, fop = 65535, rip = 4294967295, rdp = 0, mxcsr = 8064, mxcr_mask = 0, _st = {{significand = {0, 0, 0, 0}, exponent = 0, 
                __glibc_reserved1 = {0, 0, 0}}, {significand = {0, 0, 0, 0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}, {significand = {0, 0, 0, 0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}, {significand = {0, 0, 0, 
                  0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}, {significand = {0, 0, 0, 0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}, {significand = {0, 0, 0, 0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}, {
                significand = {0, 0, 0, 0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}, {significand = {0, 0, 0, 0}, exponent = 0, __glibc_reserved1 = {0, 0, 0}}}, _xmm = {{element = {0, 0, 0, 0}} <repeats 14 times>, {
                element = {1939654848, 32599, 3850395952, 32767}}, {element = {1944465392, 32599, 3850396400, 32767}}}, __glibc_reserved1 = {1948692024, 32599, 3850395952, 32767, 3850396352, 32767, 1944465392, 32599, 
              1944465392, 32599, 3850396480, 32767, 1948768059, 32599, 0, 16777216, 1, 0, 1948843952, 32599, 1944462400, 32599, 0, 0}}, __ssp = {1948764751, 8, 140013583344704, 140737043784544}}
        old_fcw = 895
#10 0x00007f5774287420 in GC_do_blocking (fn=0x208970 <reverse_test_inner>, client_data=0x0) at extra/../misc.c:2366
        my_data = {fn = 0x208970 <reverse_test_inner>, client_data = 0x0}
#11 0x00000000002080a4 in reverse_test () at tests/gctest.c:996
No locals.
#12 0x00000000002066e1 in run_one_test () at tests/gctest.c:1911
        ws = L"abc"
        x = 0x7f577401bff0 "a"
        y = 0x206d60 <fail_proc1> "UH\211\345H\211}\370H\213\005\361\063\006"
        z = 0x7f5773f54f50
        start_time = {tv_sec = 1129619, tv_nsec = 418266725}
        reverse_time = {tv_sec = 0, tv_nsec = 0}
        time_diff = 555595507
        pid = 492805
        wstatus = 0
--Type <RET> for more, q to quit, c to continue without paging--
        thr_hndl_sb = {gc_thread_handle = 0x7f5774362000 <GC_stackbottom>, sb = {mem_base = 0x7fffe5806720}}
#13 0x00000000002052c8 in main () at tests/gctest.c:2325
        i = 2
(gdb) 
```
</details>

I'm pretty sure I've gotten the compilation flags etc correct by now so I'm really quite stumped by this and I'm not sure what to do.

ivmai commented:
Please create a separate issue about the failure.

plajjan commented:
I discovered one more `-Werror`, not entirely sure why it doesn't show up in the CMake builds since it also applies `-Werror` but I suspect this is zigs extra default flags to clang to enable more sanitization and similar.

plajjan commented:
Ok, I addressed the warning / error in mallo.c and enabled `-Werror` in last commit, hope you're ok with that sort of fix!?

plajjan commented:
There are actually 2 tests out of the 12 that intermittently fail and we can see it failing in this last run after my latest commit, see https://github.com/ivmai/bdwgc/actions/runs/7281080470?pr=598

plajjan commented:
@ivmai Apart from this intermittently failing test I think I've addressed all your feedback and improved various other things along the way. I don't think I have any more changes planned, so I think another review of build.zig would be in place :)

ivmai commented:
I have push your change of readme with some modification. Plus 2 your changes.

ivmai commented on build.zig:
```diff
@@ -0,0 +1,459 @@
+    // We HAVE_WCSLEN, so enabling..
+    flags.append("-D GC_REQUIRE_WCSDUP") catch unreachable;
+
+    if (optimize == .Debug) {
```
Should be, I guess: optimize != .Debug

ivmai commented on build.zig:
```diff
@@ -0,0 +1,459 @@
+
+    if (build_shared_libs) {
+        flags.append("-D GC_DLL") catch unreachable;
+// TODO: what to do about visibility=hidden? is it good or bad?
```
I think clang supports it everywhere (it might be meaningless on some targets, like Windows DLL but should not be harmful)

ivmai commented on build.zig:
```diff
@@ -0,0 +1,459 @@
+        "-fno-strict-aliasing",
+        "-O2",
+    }) catch unreachable;
+    // TODO: why is -O2 above necessary? without it, i.e. -O0, we get
```
Please create an issue regarding gctest fails if -O0

ivmai commented on build.zig:
```diff
@@ -0,0 +1,459 @@
+    }) catch unreachable;
+
+    var have_getcontext = true;
+    if (t.abi.isMusl()) {
```
Unavailability of feature checking is, of course, a bad thing of Zig or at least of our initial version of the script.

plajjan replied:
Agreed. Not sure what the idiomatic way is. I want to look into that more but I don't think it should block this PR... We can always improve later on, right?

ivmai replied:
Right. I hope today I will perform testing and final review.

ivmai commented on .github/workflows/zig-xbuild.yml:
```diff
@@ -0,0 +1,29 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        ttriple: [ aarch64-linux-musl, wasm32-wasi, x86_64-linux-gnu.2.27, x86_64-linux-musl, x86_64-windows-gnu ]
```
How to cross-compile for MacOS? I've failed to

plajjan replied:
Should be a simple matter of: `zig build -Dtarget=x86_64-macos-none`

Or aarch64-macos-none for apple silicon...

ivmai replied:
This results in an error:
`.../bdwgc/extra/../os_dep.c:4354:10: error: 'mach/exception.h' file not found`

ivmai replied:
E.g. see https://github.com/ivmai/bdwgc/actions/runs/7323413734/job/19945962265

plajjan replied:
@ivmai ah yes, that's correct. I forgot I'm running with a patched zig. Zig doesn't ship with the correct headers just yet. I opened an issue for this https://github.com/ziglang/zig/issues/18257. They're usually quite fast adding headers but I'm guessing it'll be after the holidays.

In Acton we add in a few extra headers, see https://github.com/actonlang/acton/tree/main/deps/zig-extras/lib/libc/include/any-macos-any

You can do the same, just copy into your zig/lib/...

ivmai replied:
Okay, let's wait for right nightly build and then add macos targets.

plajjan replied:
Right yes, agreed, let's wait for it.. I was just thinking if you wanted to test stuff locally first :)

ivmai commented:
I've pushed your PR with plenty of modifications. Please test it.
I'm still processing zig-build.yml

ivmai commented:
> Please create a separate issue about the failure.

I've created #601
